File: nudge-next/src/features/whatsapp/outbox.ts
/**
 * Transactional outbound outbox — a port of `backend/tools/outbox.py`
 * (`enqueue_message` + `process_outbox`) over the `WhatsAppRepo` port.
 *
 * The AI graph / opt-out / clarify flows never call Meta as a side effect; they
 * enqueue a durable row and this worker drains it with retry + backoff, marking
 * rows `dead` after `maxAttempts`. The `$id = sha256(workspaceId:idempotencyKey)`
 * collision (uq_outbox_idem) guarantees the same logical reply is enqueued at
 * most once, so duplicate webhooks / retries can never double-send.
 *
 * Backoff is VERBATIM: on failure `attempts++`; `dead` at `maxAttempts`, else
 * `failed` with `nextAttemptAt = now + min(300, 2**attempts)s`.
 *
 * SENDS/DAY CAP (Phase 7, §6 'critical'): the per-day send limit is enforced HERE
 * in the async worker — not just at campaign-create — via an optional
 * `remainingSends(workspaceId)` resolver. Before sending a row, the worker asks
 * the resolver how many sends that workspace has left today; when a workspace's
 * budget is exhausted, its remaining due rows are skipped (left `pending`/`failed`
 * for the next drain after midnight). The resolver DEFAULTS to Infinity, so every
 * existing test (which passes none) is unaffected.
 */
import type { MessageSender } from "./send";
import { sendText } from "./send";
import type { EnqueueOutboxInput, OutboxRow, WhatsAppRepo } from "./repo";

export const MAX_BACKOFF_SECONDS = 300;

/**
 * Enqueue an outbound message. Idempotent on (workspaceId, idempotencyKey):
 * a second call with the same key returns the existing row (never a duplicate).
 */
export async function enqueueMessage(
  repo: WhatsAppRepo,
  input: EnqueueOutboxInput,
): Promise<OutboxRow> {
  const created = await repo.enqueueOutboxOrNull(input);
  if (created) return created;
  // Lost the race / already enqueued — return the existing row.
  const existing = await repo.getOutboxByKey(
    input.workspaceId,
    input.idempotencyKey,
  );
  if (!existing) {
    // Should be unreachable (create said duplicate but row is gone); surface it.
    throw new Error(
      `enqueueMessage: outbox row vanished for ${input.workspaceId}:${input.idempotencyKey}`,
    );
  }
  return existing;
}

export interface DrainStats {
  processed: number;
  sent: number;
  failed: number;
  dead: number;
}

export interface DrainOptions {
  limit?: number;
  now?: Date;
  /** Injectable sender (parity with the Python `send_fn`). Defaults to sendText. */
  sender?: MessageSender;
  /**
   * Per-workspace remaining sends today (the plan §6 cap). Returns how many more
   * messages `workspaceId` may send today; rows are skipped once its budget hits
   * 0. DEFAULTS to Infinity (no cap) so existing callers/tests are unaffected.
   * Resolved at most once per workspace per drain.
   */
  remainingSends?: (workspaceId: string) => Promise<number> | number;
}

/**
 * Drain due rows once. Claims `status∈{pending,failed}`, `attempts<maxAttempts`,
 * due (`nextAttemptAt` null or <= now), oldest-first up to `limit`. For each:
 * optimistic `pending/failed → sending` + `attempts++` (Appwrite has no SELECT
 * FOR UPDATE), send, then:
 *   - success ⇒ `sent` + providerMessageId + sentAt; backfill inbox wamid.
 *   - failure ⇒ `dead` at maxAttempts, else `failed` with backoff nextAttemptAt.
 */
export async function drainOutbox(
  repo: WhatsAppRepo,
  options: DrainOptions = {},
): Promise<DrainStats> {
  const now = options.now ?? new Date();
  const limit = options.limit ?? 25;
  const send = options.sender ?? sendText;
  const remainingSendsFor =
    options.remainingSends ?? (() => Number.POSITIVE_INFINITY);

  const rows = await repo.listDueOutbox(now, limit);
  const stats: DrainStats = { processed: 0, sent: 0, failed: 0, dead: 0 };

  // Per-workspace remaining-send budget for this drain (resolved once each).
  const budget = new Map<string, number>();
  async function budgetFor(workspaceId: string): Promise<number> {
    const cached = budget.get(workspaceId);
    if (cached !== undefined) return cached;
    const remaining = await remainingSendsFor(workspaceId);
    budget.set(workspaceId, remaining);
    return remaining;
  }

  for (const row of rows) {
    // Sends/day cap: skip this row if its workspace has no budget left today.
    // The row stays claimable for the next drain (after UTC midnight resets).
    if ((await budgetFor(row.workspaceId)) <= 0) {
      continue;
    }

    stats.processed += 1;
    const attempts = row.attempts + 1;
    // Optimistic claim: flip to sending + bump attempts before the send.
    await repo.updateOutbox(row.$id, { status: "sending", attempts });

    let payload: { to?: string; text?: string; sender?: string | null };
    try {
      payload = JSON.parse(row.payloadJson ?? "{}");
    } catch {
      payload = {};
    }

    try {
      const pmid = await send(
        payload.to ?? "",
        payload.text ?? "",
        payload.sender ?? null,
      );
      await repo.updateOutbox(row.$id, {
        status: "sent",
        providerMessageId: pmid,
        sentAt: new Date().toISOString(),
        lastError: null,
      });
      // Backfill the outbound inbox message's wamid (only if not already set).
      if (row.inboxMessageId && pmid) {
        await repo.setInboxMessageWamid(row.inboxMessageId, pmid);
      }
      // Consume one unit of the workspace's daily send budget.
      budget.set(row.workspaceId, (budget.get(row.workspaceId) ?? 0) - 1);
      stats.sent += 1;
    } catch (err) {
      const lastError = (err instanceof Error ? err.message : String(err)).slice(
        0,
        1000,
      );
      if (attempts >= row.maxAttempts) {
        await repo.updateOutbox(row.$id, { status: "dead", lastError });
        stats.dead += 1;
      } else {
        const backoff = Math.min(MAX_BACKOFF_SECONDS, 2 ** attempts);
        const nextAttemptAt = new Date(now.getTime() + backoff * 1000).toISOString();
        await repo.updateOutbox(row.$id, {
          status: "failed",
          nextAttemptAt,
          lastError,
        });
        stats.failed += 1;
      }
    }
  }

  return stats;
}

File: backend/tools/outbox.py
"""
Transactional outbound message outbox.

The AI graph (and opt-out / clarification flows) never call the WhatsApp provider
as a side effect. They `enqueue_message(...)` a durable row; a worker drains the
queue with retry + backoff and marks rows `dead` after `max_attempts`.

The (workspace_id, idempotency_key) unique constraint guarantees the same logical
reply is enqueued at most once, so duplicate webhooks / retries can never cause
duplicate sends.
"""
from __future__ import annotations

import logging
from datetime import datetime, timedelta

from sqlalchemy.exc import IntegrityError

from db.models import InboxMessage, MessageOutbox
from tools import wa_provider

logger = logging.getLogger(__name__)

MAX_BACKOFF_SECONDS = 300


def enqueue_message(db, *, workspace_id, to, text, sender=None,
                    conversation_id=None, contact_id=None, inbox_message_id=None,
                    idempotency_key, channel="whatsapp") -> MessageOutbox:
    """Enqueue an outbound message. Idempotent on (workspace_id, idempotency_key):
    a second call with the same key returns the existing row, never a duplicate."""
    existing = (db.query(MessageOutbox)
                .filter(MessageOutbox.workspace_id == workspace_id,
                        MessageOutbox.idempotency_key == idempotency_key).first())
    if existing:
        return existing

    row = MessageOutbox(
        workspace_id=workspace_id, contact_id=contact_id, conversation_id=conversation_id,
        inbox_message_id=inbox_message_id, channel=channel,
        payload={"to": to, "text": text, "sender": sender},
        idempotency_key=idempotency_key, status="pending", attempts=0,
        next_attempt_at=None,
    )
    try:
        with db.begin_nested():   # SAVEPOINT — a unique clash rolls back only this insert
            db.add(row)
        return row
    except IntegrityError:
        # Lost a race — the row now exists; return it.
        return (db.query(MessageOutbox)
                .filter(MessageOutbox.workspace_id == workspace_id,
                        MessageOutbox.idempotency_key == idempotency_key).first())


def process_outbox(db, *, limit: int = 25, now: datetime | None = None, send_fn=None) -> dict:
    """Drain due rows once. `now` and `send_fn` are injectable for tests."""
    now = now or datetime.utcnow()
    send_fn = send_fn or wa_provider.send_text
    rows = (db.query(MessageOutbox)
            .filter(MessageOutbox.status.in_(["pending", "failed"]),
                    MessageOutbox.attempts < MessageOutbox.max_attempts)
            .filter((MessageOutbox.next_attempt_at.is_(None)) | (MessageOutbox.next_attempt_at <= now))
            .order_by(MessageOutbox.created_at.asc())
            .limit(limit).all())

    stats = {"processed": 0, "sent": 0, "failed": 0, "dead": 0}
    for row in rows:
        stats["processed"] += 1
        row.status = "sending"
        row.attempts += 1
        db.commit()
        try:
            pmid = send_fn(row.payload.get("to"), row.payload.get("text"), row.payload.get("sender"))
            row.status = "sent"
            row.provider_message_id = pmid
            row.sent_at = datetime.utcnow()
            row.last_error = None
            if row.inbox_message_id and pmid:
                msg = db.query(InboxMessage).filter(InboxMessage.id == row.inbox_message_id).first()
                if msg and not msg.wamid:
                    msg.wamid = pmid
            stats["sent"] += 1
        except Exception as exc:
            row.last_error = str(exc)[:1000]
            if row.attempts >= row.max_attempts:
                row.status = "dead"
                stats["dead"] += 1
                logger.warning("[outbox] message %s dead after %s attempts: %s", row.id, row.attempts, exc)
            else:
                row.status = "failed"
                row.next_attempt_at = now + timedelta(seconds=min(MAX_BACKOFF_SECONDS, 2 ** row.attempts))
                stats["failed"] += 1
        db.commit()
    return stats
